src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1284 @@
+                var labelsToIgnore = PooledHashSet<LabelSymbol>.GetInstance();
+                populateStateForCases(builder, rootIdentifier, previousCases, patternIndex, orCases, labelsToIgnore, syntax, ref casesBuilder.AsRef());
+                BoundDecisionDag dag = builder.MakeBoundDecisionDag(syntax, ref casesBuilder.AsRef());
+
+                foreach (StateForCase @case in casesBuilder)
+                {
+                    if (!dag.ReachableLabels.Contains(@case.CaseLabel) && !labelsToIgnore.Contains(@case.CaseLabel))
+                    {
+                        ErrorCode errorCode = detectNotOrPattern(@case.Syntax) ? ErrorCode.WRN_RedundantPattern : ErrorCode.HDN_RedundantPattern;
+                        diagnostics.Add(errorCode, @case.Syntax);
+                    }
+                }
+
+                labelsToIgnore.Free();
+            }
+
+            return;
+
+            static void populateStateForCases(DecisionDagBuilder builder, BoundDagTemp rootIdentifier, ArrayBuilder<StateForCase> previousCases, int patternIndex,
+                OrCases set, PooledHashSet<LabelSymbol> labelsToIgnore, SyntaxNode nodeSyntax, ref TemporaryArray<StateForCase> casesBuilder)
+            {
+                for (int i = 0; i < patternIndex; i++)
+                {
+                    casesBuilder.Add(previousCases[i]);
+                }
+
+                int index = patternIndex;
+                foreach ((BoundPattern pattern, SyntaxNode? syntax) in set.Cases)
+                {
+                    var label = new GeneratedLabelSymbol("orCase");
+                    SyntaxNode? diagSyntax = syntax;
+                    if (diagSyntax is null)
+                    {
+                        labelsToIgnore.Add(label);
+                        diagSyntax = nodeSyntax;
+                    }
+
+                    Debug.Assert(diagSyntax is not null);
+                    casesBuilder.Add(builder.MakeTestsForPattern(++index, diagSyntax, rootIdentifier, pattern, whenClause: null, label: label));
+                }
+            }
+
+            static bool detectNotOrPattern(SyntaxNode syntax)
+            {
+                // We need to reduce the break introduce by flagging redundant patterns
+                // and we never want to affect people who express their patterns thoroughly (but correctly)
+                // such as `switch { < 0 => -1, 0 => 0, > 0 => 1 }`
+                // So we're only reporting a warning for situations that syntactically look hazardous
+                // At the moment, we're only interested in patterns involved in a `not ... or ...` pattern
+
+                // If the pattern is on the right of an `or` pattern, we walk up and
+                // if any of the preceding patterns is a `not` pattern we've detected the error-prone not/or situation.
+                // For example: `not A or <redundant>` or `A or not B or <redundant>`
```

Comment:
>// For example: `not A or <redundant>` or `A or not B or <redundant>` [](http://example.com/codeflow?start=16&length=69)

It might be worth explicitly stating whether we are interested detecting ```not A or (<redundant> or/and ... )```

Reply:
This thread was marked as resolved, but it doesn't look like the desired behavior for the scenario(s) was clarified